Engine/Helper.cs

```diff
@@ -602,73 +602,50 @@ public bool HasSplattedVariable(CommandAst cmdAst)
         }
 
         /// <summary>
-        /// Given a commandast, checks whether positional parameters are used or not.
+        /// Given a commandast, checks if the command is a Cmdlet.
         /// </summary>
         /// <param name="cmdAst"></param>
-        /// <param name="moreThanThreePositional">only return true if more than three positional parameters are used</param>
         /// <returns></returns>
-        public bool PositionalParameterUsed(CommandAst cmdAst, bool moreThanThreePositional = false)
-        {
+        public bool IsCmdlet(CommandAst cmdAst) {
             if (cmdAst == null)
             {
                 return false;
             }
 
-            var commandInfo = GetCommandInfoLegacy(cmdAst.GetCommandName());
-            if (commandInfo == null || (commandInfo.CommandType != System.Management.Automation.CommandTypes.Cmdlet))
-            {
-                return false;
-            }
-
-            IEnumerable<ParameterMetadata> switchParams = null;
+            var commandInfo = GetCommandInfo(cmdAst.GetCommandName());
+            return (commandInfo != null && commandInfo.CommandType == System.Management.Automation.CommandTypes.Cmdlet);
+        }
 
+        /// <summary>
+        /// Given a commandast, checks whether positional parameters are used or not.
+        /// </summary>
+        /// <param name="cmdAst"></param>
+        /// <param name="moreThanTwoPositional">only return true if more than three positional parameters are used</param>
+        /// <returns></returns>
+        public bool PositionalParameterUsed(CommandAst cmdAst, bool moreThanTwoPositional = false)
+        {
             if (HasSplattedVariable(cmdAst))
             {
                 return false;
             }
 
-            if (commandInfo != null && commandInfo.CommandType == System.Management.Automation.CommandTypes.Cmdlet)
-            {
-                try
-                {
-                    switchParams = commandInfo.Parameters.Values.Where<ParameterMetadata>(pm => pm.SwitchParameter);
-                }
-                catch (Exception)
-                {
-                    switchParams = null;
-                }
-            }
-
-            int parameters = 0;
             // Because of the way we count, we will also count the cmdlet as an argument so we have to -1
-            int arguments = -1;
+            int argumentsWithoutProcedingParameters = -1;
+            bool parameterPreceding = false;
 
             foreach (CommandElementAst ceAst in cmdAst.CommandElements)
```

Comment:
I think this can all be written more simply:
```
var commandElementCollection = cmdAst.CommandElements;
if ( commandElementCollection.Count = 1 ) { return false; }
for(int i = 1; i < commandElementCollection.Length; i++) {
if ( commandElementCollection[i] isnot CommandParameterAst && commandElementCollection[i-1] isnot CommandParameterAst ) {
argumentsWithoutProcedingParameters++;
}
}
```
this should catch things like `test-cmd a -a -b c,f d,z $a $b -q $() -e -f @() @{ one = 1 } -z "this is a test" "this is a test"`
it doesn't validate that there may be missing parameter arguments, but it should catch any positional parameters

Reply:
Fixed.